### ISSUE TITLE
[Beta: CodeMirror] User can undo past the current challenge

### ISSUE THREAD

**Issue by leonfeng** [freeCodeCamp/freeCodeCamp#16960]
#### Challenge Name

Any challenge with a code editor.

#### Issue Description

Steps to reproduce:
1. Complete a challenge.
2. Enter some code in the new challenge.
3. Keep undoing typing and your current code will be reverted to that of the previous challenge.

#### Browser Information

* Browser Name, Version: Chrome 65.0.3325.181
* Operating System: macOS 10.13.3
* Mobile, Desktop, or Tablet: desktop

#### Screenshot

![freecodecamp beta - undo issue](https://user-images.githubusercontent.com/241553/37878376-8ad726ca-3079-11e8-8ee5-e6eb86bf705c.gif)

**Comment by love1024** [comment]
@raisedadead  we need to call clearHistory of codemirror to get rid of undo history when new challenge comes. But when we know that this is new challenge so that we can clear history in Editor.jsx? componentWillMount and componentWillUpdate does not seem right for this. Any Suggestion?

**Comment by raisedadead** [comment]
You should be able to hook it to an action in the editor, when a new challenge arrives. /cc @Bouncey can you guide.

**Comment by love1024** [comment]
Yes i am getting id of challenge to compare if this challenge is new in componentWillUpdate and then clearing history. This is working fine expect for the first time.  Working on this.

**Comment by love1024** [comment]
@raisedadead @Bouncey   There are two ways to do this \
First is to add setTimeout to clear codemirror history 
```
  componentDidUpdate(nextProps) {
    if (nextProps.id !== this.props.id) {
      setTimeout(() => {
        this.refs.editor.getCodeMirror().doc.clearHistory();
      }, 0);
    }
  }
```
Settimeout is required because codemirror does not give updated value when new challenge is selected as of this issue https://github.com/JedWatson/react-codemirror/issues/106 \
Second is to use the workaround given in issue to add key to codemirror component to get updated value.I used id of the challenge for this
```
<codemirror
key = {id}
...
/>
```
So Any Suggestion which one to use?

**Comment by raisedadead** [comment]
I would want to move away to using https://github.com/scniro/react-codemirror2 however, in case its not trivial at the moment I would prefer your second solution.

**Comment by love1024** [comment]
yes codemirror2 seems a right option, for now i am raising PR with second option.
